src/core/thread/dua_manager.cpp

```diff
@@ -184,6 +241,501 @@ otError DuaManager::Store(void)
+    Ip6::MessageInfo messageInfo;
+    Ip6::Address     dua;
+
+    VerifyOrExit(mle.IsAttached(), error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(Get<BackboneRouter::Leader>().HasPrimary(), error = OT_ERROR_INVALID_STATE);
+
+    // Only allow one outgoing DUA.req
+    VerifyOrExit(!mIsDuaPending, error = OT_ERROR_BUSY);
+
+    // Only send DUA.req when necessary
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+    VerifyOrExit(mle.IsRouterOrLeader() || !mle.IsExpectedToBecomeRouter(), error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit((mDuaState == DuaState::kToRegister && mDelay.mFields.mRegistrationDelay == 0) ||
+                     mChildDuaMask != mChildDuaRegisteredMask,
+                 error = OT_ERROR_NOT_FOUND);
+#else
+    VerifyOrExit(mDuaState == DuaState::kToRegister && mDelay.mFields.mRegistrationDelay == 0,
+                 error = OT_ERROR_NOT_FOUND);
+#endif // OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+
+    VerifyOrExit(mle.IsFullThreadDevice() || mle.GetParent().IsThreadVersion1p1(), error = OT_ERROR_INVALID_STATE);
+#endif // OPENTHREAD_CONFIG_DUA_ENABLE
+
+    // Prepare DUA.req
+    VerifyOrExit((message = Get<Coap::Coap>().NewPriorityMessage()) != nullptr, error = OT_ERROR_NO_BUFS);
+
+    SuccessOrExit(error =
+                      message->Init(OT_COAP_TYPE_CONFIRMABLE, OT_COAP_CODE_POST, OT_URI_PATH_DUA_REGISTRATION_REQUEST));
+    SuccessOrExit(error = message->SetPayloadMarker());
+
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+    if (mDuaState == DuaState::kToRegister && mDelay.mFields.mRegistrationDelay == 0)
+    {
+        dua = GetDomainUnicastAddress();
+        SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kTarget, &dua, sizeof(dua)));
+        SuccessOrExit(error = Tlv::AppendTlv(*message, ThreadTlv::kMeshLocalEid, &mle.GetMeshLocal64().GetIid(),
+                                             sizeof(Ip6::InterfaceIdentifier)));
+        mDuaState             = DuaState::kRegistering;
+        mLastRegistrationTime = TimerMilli::GetNow();
+    }
+    else
+#endif // OPENTHREAD_CONFIG_DUA_ENABLE
+    {
+#if OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
+        uint32_t            lastTransactionTime;
+        const Ip6::Address *duaPtr = nullptr;
+        Child *             child  = nullptr;
+
+        if (!mRegisterCurrentChildIndex)
+        {
+            for (Child &iter: Get<ChildTable>().Iterate(Child::kInStateValid))
+            {
+                uint16_t childIndex = Get<ChildTable>().GetChildIndex(iter);
+
+                if (mChildDuaMask.Get(childIndex) && !mChildDuaRegisteredMask.Get(childIndex))
+                {
+                    mChildIndexDuaRegistering = childIndex;
+                    break;
+                }
+            }
+        }
+
+        child = Get<ChildTable>().GetChildAtIndex(mChildIndexDuaRegistering);
+
+        VerifyOrExit((duaPtr = child->GetDomainUnicastAddress()) != nullptr, error = OT_ERROR_ABORT);
```

Comment:
```suggestion
        VerifyOrExit(IsChildHasDuaToRegister(mChildIndexDuaRegistering), error = OT_ERROR_NOT_FOUND);
```
Maybe we should also check the two masks, mChildDuaMask and `mChildDuaRegisteredMask` ? 

Also, my personal suggestion for the error in such case is `OT_ERROR_NOT_FOUND`. Thoughts?

Reply:
It is already checked at the beginning
 >  VerifyOrExit((mDuaState == kToRegister && mDelay.mFields.mRegistrationDelay == 0) ||
                       mChildDuaMask != mChildDuaRegisteredMask, error = OT_ERROR_NOT_FOUND);

Will update it as below for enhancement
> VerifyOrExit((mDuaState == kToRegister && mDelay.mFields.mRegistrationDelay == 0) ||
                       (mChildDuaMask.HasAny() && mChildDuaMask != mChildDuaRegisteredMask), error = OT_ERROR_NOT_FOUND);

Any further concerns?

Reply:
Maybe better to  use assert as the condition checks at the beginning already ensured there is child that has DUA to register here.

Reply:
I would prefer to be more strict in general.